store.go

```diff
@@ -0,0 +1,195 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/tidepool-org/go-common/clients/mongo"
+	"labix.org/v2/mgo"
+	"labix.org/v2/mgo/bson"
+)
+
+const (
+	data_collection = "deviceData"
+)
+
+type (
+	//Interface for the query iterator
+	StorageIterator interface {
+		Next(result interface{}) bool
+		Close() error
+	}
+	//Interface for our storage layer
+	Storage interface {
+		Close()
+		Ping() error
+		GetDeviceData(p *params) StorageIterator
+	}
+	//Mongo Storage Client
+	MongoStoreClient struct {
+		session *mgo.Session
+	}
+
+	params struct {
+		active bool
+		//userId comes from the request
+		userId string
+		//groupId is resolved from the incoming userid and if used storage in queries
+		groupId  string
+		types    []string
+		subTypes []string
+		date
+		schemaVersion *SchemaVersion
+	}
+
+	date struct {
+		start string
+		end   string
+	}
+)
+
+func getParams(q url.Values, schema *SchemaVersion) (*params, error) {
+
+	startStr := q.Get("startdate")
+	endStr := q.Get("enddate")
+
+	if startStr != "" {
+		startDate, err := time.Parse(time.RFC3339Nano, startStr)
+		if err != nil {
+			return nil, err
+		}
+		startStr = startDate.Format(time.RFC3339Nano)
+	}
+	if endStr != "" {
+		endDate, err := time.Parse(time.RFC3339Nano, endStr)
+		if err != nil {
+			return nil, err
+		}
+		endStr = endDate.Format(time.RFC3339Nano)
+	}
+
+	p := &params{
+		userId: q.Get(":userID"),
+		active: true,
+		//the query params for type and subtype can contain multiple values seperated by a comma e.g. "type=smbg,cbg"
+		//so split them out into an array of values
+		types:         strings.Split(q.Get("type"), ","),
+		subTypes:      strings.Split(q.Get("subtype"), ","),
+		date:          date{startStr, endStr},
+		schemaVersion: schema,
+	}
+
+	log.Println(DATA_API_PREFIX, fmt.Sprintf("****Params: %v", p))
+
+	return p, nil
+
+}
+
+func NewMongoStoreClient(config *mongo.Config) *MongoStoreClient {
+
+	mongoSession, err := mongo.Connect(config)
+	if err != nil {
+		log.Fatal(DATA_API_PREFIX, err)
+	}
+
+	//index based on sort and where keys
+	index := mgo.Index{
+		Key:        []string{"_groupId", "_active", "_schemaVersion"},
+		Background: true,
+	}
+	err = mgoDataCollection(mongoSession).EnsureIndex(index)
```

Comment:
I see three instances of `mgoDataCollection(mongoSession)`. Should there just be one?